nano/node/ipc.cpp

```diff
@@ -0,0 +1,682 @@
+	 */
+	boost::asio::io_context & io_ctx;
+
+	/** A socket of the given asio type */
+	SOCKET_TYPE socket;
+
+	/** Buffer sizes are read into this */
+	uint32_t buffer_size{ 0 };
+
+	/** Buffer used to store data received from the client */
+	std::vector<uint8_t> buffer;
+
+	/** Transport configuration */
+	nano::ipc::ipc_config_transport & config_transport;
+};
+
+/** Domain and TCP socket transport */
+template <typename ACCEPTOR_TYPE, typename SOCKET_TYPE, typename ENDPOINT_TYPE>
+class socket_transport : public nano::ipc::transport
+{
+public:
+	socket_transport (nano::ipc::ipc_server & server_a, ENDPOINT_TYPE endpoint_a, nano::ipc::ipc_config_transport & config_transport_a, int concurrency_a) :
+	server (server_a), config_transport (config_transport_a)
+	{
+		// Using a per-transport event dispatcher?
+		if (concurrency_a > 0)
+		{
+			io_ctx = std::make_unique<boost::asio::io_context> ();
+		}
+
+		boost::asio::socket_base::reuse_address option (true);
+		boost::asio::socket_base::keep_alive option_keepalive (true);
+		acceptor = std::make_unique<ACCEPTOR_TYPE> (context (), endpoint_a);
+		acceptor->set_option (option);
+		acceptor->set_option (option_keepalive);
+		accept ();
+
+		// Start serving IO requests. If concurrency_a is < 1, the node's thread pool/io_context is used instead.
+		// A separate io_context for domain sockets may facilitate better performance on some systems.
+		if (concurrency_a > 0)
+		{
+			runner = std::make_unique<nano::thread_runner> (*io_ctx, concurrency_a);
+		}
+	}
+
+	boost::asio::io_context & context () const
+	{
+		return io_ctx ? *io_ctx : server.node.io_ctx;
+	}
+
+	void accept ()
+	{
+		// Prepare the next session
+		auto new_session (std::make_shared<session<SOCKET_TYPE>> (server, context (), config_transport));
+
+		acceptor->async_accept (new_session->get_socket (), [this, new_session](boost::system::error_code const & ec) {
+			if (!ec)
+			{
+				new_session->read_next_request ();
+			}
+			else
+			{
+				BOOST_LOG (server.node.log) << "IPC: acceptor error: " << ec.message ();
+			}
+
+			if (acceptor->is_open () && ec != boost::asio::error::operation_aborted)
+			{
+				this->accept ();
+			}
+			else
+			{
+				BOOST_LOG (server.node.log) << "IPC: shutting down";
+			}
+		});
+	}
+
+	void stop ()
+	{
+		acceptor->close ();
+		if (io_ctx)
+		{
+			io_ctx->stop ();
+		}
+
+		if (runner)
+		{
+			runner->join ();
+		}
+	}
+
+private:
+	nano::ipc::ipc_server & server;
+	nano::ipc::ipc_config_transport & config_transport;
+	std::unique_ptr<nano::thread_runner> runner;
+	std::unique_ptr<boost::asio::io_context> io_ctx;
+	std::unique_ptr<ACCEPTOR_TYPE> acceptor;
+};
+
+/** The domain socket file is attemped removed at both startup and shutdown. */
+class nano::ipc::dsock_file_remover
+{
+public:
+	dsock_file_remover (std::string const & file_a) :
+	filename (file_a)
+	{
+		std::remove (filename.c_str ());
+	}
+	~dsock_file_remover ()
```

Comment:
Either make the class `final` (preventing subclassing) or make the destructor virtual